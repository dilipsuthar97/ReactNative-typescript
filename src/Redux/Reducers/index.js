import MainReducer from './MainReducer';

let rootReducer = {
    Main: MainReducer
}

export default rootReducer;