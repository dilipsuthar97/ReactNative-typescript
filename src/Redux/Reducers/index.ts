import MainReducer from './MainReducer';

const rootReducer = {
    Main: MainReducer
}

export default rootReducer;