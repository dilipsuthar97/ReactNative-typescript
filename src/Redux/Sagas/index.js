import mainSaga from './MainSaga';

const rootSaga = function* rootSaga() {
    yield* mainSaga()
}

export default rootSaga;