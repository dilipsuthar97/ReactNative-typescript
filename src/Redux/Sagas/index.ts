import { all, fork } from 'redux-saga/effects';
import mainSaga from './MainSaga';

const rootSaga = function* rootSaga() {
	// yield* mainSaga()
	yield all([fork(mainSaga)]);
};

export default rootSaga;
