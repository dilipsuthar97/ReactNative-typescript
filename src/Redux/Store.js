// >>>>>>> LIBRARIES
import { applyMiddleware, createStore, compose } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { persistStore, persistCombineReducers } from 'redux-persist';
import logger from 'redux-logger';
import AsyncStorage from '@react-native-community/async-storage';

// >>>>>>> ASSETS
import rootReducer from './Reducers';
import rootSaga from './Sagas';

const reducers = persistCombineReducers({
    key: "root",
    storage: AsyncStorage
}, rootReducer);

const sagaMiddleware = createSagaMiddleware();
const middlewares = [];
const enhancers = [];
middlewares.push(sagaMiddleware, logger);
enhancers.push(applyMiddleware(...middlewares));

export const store = createStore(
    reducers,
    compose(...enhancers),
);

// >>>>>>> persistStore contains all the data from store
export const persistor = persistStore(store);
sagaMiddleware.run(rootSaga);