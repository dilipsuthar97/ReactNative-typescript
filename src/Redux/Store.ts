// >>>>>>> LIBRARIES
import { applyMiddleware, createStore, compose } from 'redux';
import createSagaMiddleware, { SagaMiddleware } from 'redux-saga';
import { persistStore, persistCombineReducers } from 'redux-persist';
import logger from 'redux-logger';
import AsyncStorage from '@react-native-community/async-storage';

// compose dev tools
import { composeWithDevTools } from 'redux-devtools-extension';

// >>>>>>> ASSETS
import rootReducer from './Reducers';
import rootSaga from './Sagas';

const reducers = persistCombineReducers(
	{
		key: 'root',
		storage: AsyncStorage,
	},
	rootReducer
);

const sagaMiddleware = createSagaMiddleware();
const middlewares = [];
const enhancers = [];
middlewares.push(sagaMiddleware, logger);
enhancers.push(applyMiddleware(...middlewares));

// type/interface
export type AppState = ReturnType<typeof reducers>;

export const store = createStore(reducers, composeWithDevTools(...enhancers));

// >>>>>>> persistStore contains all the data from store
export const persistor = persistStore(store);
sagaMiddleware.run(rootSaga);
