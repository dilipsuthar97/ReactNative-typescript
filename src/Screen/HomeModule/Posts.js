// =======>>>>>>>> LIBRARIES <<<<<<<<=======
import React from 'react'
import { StyleSheet, View, Text, TouchableOpacity, Image, ActivityIndicator, FlatList } from 'react-native';
import { connect } from 'react-redux';

// =======>>>>>>>> ASSETS <<<<<<<<=======
import { AppStyle, Images, Scale, Colors } from '../../CommonConfig';
import { getPostsRequest } from '../../Redux/Actions';
import ItemPost from '../../Components/Item/ItemPost';

// =======>>>>>>>> CLASS DECLARATION <<<<<<<<=======
class Posts extends React.Component {
    // =======>>>>>>> STATE DECLARATION <<<<<<<=======
    state = {
        isLoading: false
    };

    // =======>>>>>>> LIFECYCLE METHODS <<<<<<<=======
    componentDidMount() {
        this.initHeader()

        this.setState({ isLoading: true })
        this.props.getPostsRequest();
    }

    UNSAFE_componentWillReceiveProps(nextProps) {
        if (nextProps.Main.getPostsSuccess && this.state.isLoading) {
            this.setState({ isLoading: false })
        } else if (nextProps.Main.getPostsFail && this.state.isLoading) {
            this.setState({ isLoading: false })
        }
    }

    // =======>>>>>>> METHODS INITIALIZE <<<<<<<=======
    initHeader() {
        this.props.navigation.setOptions({
            headerStyle: AppStyle.headerStyle,
            headerTitleStyle: AppStyle.headerTitleStyle,
            headerTitleAlign: 'center',
            headerLeft: ({ onPress }) => {
                // console.log('headerLeft props', props)
                return (
                    <TouchableOpacity style={{marginLeft: Scale(16)}} onPress={() => this.props.navigation.openDrawer()}>
                        <Image source={Images.IC_MENU} style={{ height: Scale(20), width: Scale(20), tintColor: Colors.WHITE }}/>
                    </TouchableOpacity>
                )
            }
        })
    }

    itemSeparator() {
        return <View style={{height: Scale(10)}}/>
    }

    renderPosts() {
        return (
            this.props.Main.data && this.props.Main.data.posts
                ? <FlatList
                    data={this.props.Main.data.posts}
                    ItemSeparatorComponent={this.itemSeparator}
                    keyExtractor={(item) => item.id.toString()}
                    renderItem={({item, index}) => <ItemPost item={item} style={{
                        marginTop: index == 0 ? Scale(10) : 0,
                        marginBottom: this.props.Main.data.posts.length - 1 == index ? Scale(10) : 0
                    }}/>}
                />
                : <View style={{justifyContent: 'center', alignItems: 'center'}}>
                    <Text>No Items!</Text>
                </View>
        )
    }

    // =======>>>>>>>> RENDER INITIALIZE <<<<<<<<=======
    render() {

        console.log('Main: ', this.props.Main.data)
        const rootStyle = this.state.isLoading ? [styles.container, {justifyContent: 'center', alignItems: 'center'}] : styles.container

        return <View style={rootStyle}>
            {this.state.isLoading ? <ActivityIndicator size="large" color={Colors.PRIMARY_COLOR}/> : this.renderPosts()}
        </View>
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: Colors.WHITE
    }
})

const mapStateToProps = state => {
    return {
        Main: state.Main
    }
}

export default connect(mapStateToProps, { getPostsRequest })(Posts);