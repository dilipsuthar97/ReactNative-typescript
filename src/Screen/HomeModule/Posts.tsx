// =======>>>>>>>> LIBRARIES <<<<<<<<=======
import * as React from 'react';
import {
	StyleSheet,
	View,
	Text,
	TouchableOpacity,
	Image,
	ActivityIndicator,
	FlatList,
} from 'react-native';
import { connect, DispatchProp } from 'react-redux';

// =======>>>>>>>> ASSETS <<<<<<<<=======
import { AppStyle, Images, Scale, Colors } from '../../CommonConfig';
import { getPostsRequest } from '../../Redux/Actions';
import ItemPost from '../../Components/Item/ItemPost';
import SafeAreaView from 'react-native-safe-area-view';

// =======>>>>>>>> TYPES <<<<<<<<=======
import { MainReducer } from '../../Types/Reducers.interface';
import { AppState } from '../../Redux/Store';
import { StackNavigationProp } from '@react-navigation/stack';
import { DrawerNavigationProp } from '@react-navigation/drawer';
import { PersistState } from 'redux-persist';
import { Post } from 'src/Types/Post.interface';

interface PostsProps {
	navigation: (StackNavigationProp<any> & DrawerNavigationProp<any>) | null;
}

interface PostsState {
	isLoading: boolean;
}

type Props = PostsProps & LinkStateProps & LinkDispatchProps;

// =======>>>>>>>> CLASS DECLARATION <<<<<<<<=======
class Posts extends React.Component<Props, PostsState> {
	// =======>>>>>>> STATE DECLARATION <<<<<<<=======
	state: PostsState = {
		isLoading: false,
	};

	// =======>>>>>>> LIFECYCLE METHODS <<<<<<<=======
	componentDidMount() {
		this.initHeader();

		// setTimeout(() => this.props.navigation.openDrawer(), 1000)

		this.setState({ isLoading: true });
		this.props.getPostsRequest();
	}

	UNSAFE_componentWillReceiveProps(nextProps: Props) {
		if (nextProps.Main.getPostsSuccess && this.state.isLoading) {
			this.setState({ isLoading: false });
		} else if (nextProps.Main.getPostsFail && this.state.isLoading) {
			this.setState({ isLoading: false });
		}
	}

	// =======>>>>>>> METHODS INITIALIZE <<<<<<<=======
	public getPosts = (): Post[] | undefined => {
		if (this.props.Main.data.posts !== undefined) {
			return this.props.Main.data.posts;
		}
	};

	initHeader() {
		const { navigation } = this.props;

		navigation !== null &&
			navigation.setOptions({
				headerStyle: AppStyle.headerStyle,
				headerTitleStyle: AppStyle.headerTitleStyle,
				headerTitleAlign: 'center',
				headerLeft: () => {
					// console.log('headerLeft props', props)
					return (
						<TouchableOpacity
							style={{ marginLeft: Scale(16) }}
							onPress={() => navigation.openDrawer()}
						>
							<Image
								source={Images.IC_MENU}
								style={{
									height: Scale(20),
									width: Scale(20),
									tintColor: Colors.WHITE,
								}}
							/>
						</TouchableOpacity>
					);
				},
			});
	}

	itemSeparator() {
		return <View style={{ height: Scale(10) }} />;
	}

	renderPosts() {
		return this.props.Main.data && this.props.Main.data.posts ? (
			<FlatList
				data={this.props.Main.data.posts}
				ItemSeparatorComponent={this.itemSeparator}
				keyExtractor={(item) => item.id.toString()}
				renderItem={({ item, index }) => (
					<ItemPost
						item={item}
						style={{
							marginTop: index == 0 ? Scale(10) : 0,
							marginBottom: this.props.Main.data.posts
								? this.props.Main.data.posts.length - 1 == index
									? Scale(10)
									: 0
								: 0,
						}}
					/>
				)}
			/>
		) : (
			<View style={{ justifyContent: 'center', alignItems: 'center' }}>
				<Text>No Items!</Text>
			</View>
		);
	}

	// =======>>>>>>>> RENDER INITIALIZE <<<<<<<<=======
	render() {
		console.log('Main: ', this.props.Main.data);
		console.log('_persist: ', this.props._persist);

		return (
			<SafeAreaView style={AppStyle.safeArea}>
				<View
					style={
						this.state.isLoading
							? {
									...styles.container,
									justifyContent: 'center',
									alignItems: 'center',
							  }
							: styles.container
					}
				>
					{this.state.isLoading ? (
						<ActivityIndicator
							size="large"
							color={Colors.PRIMARY_COLOR}
						/>
					) : (
						this.renderPosts()
					)}
				</View>
			</SafeAreaView>
		);
	}
}

const styles = StyleSheet.create({
	container: {
		flex: 1,
		backgroundColor: Colors.WHITE,
	},
});

interface LinkStateProps {
	Main: MainReducer;
	_persist: PersistState;
}

interface LinkDispatchProps {
	getPostsRequest: () => void;
}

const mapStateToProps = (
	state: AppState,
	ownProps: PostsProps
): LinkStateProps => ({
	Main: state.Main,
	_persist: state._persist,
});

const mapDispatchToProps: LinkDispatchProps = {
	getPostsRequest: () => getPostsRequest(),
};

export default connect(mapStateToProps, mapDispatchToProps)(Posts);
