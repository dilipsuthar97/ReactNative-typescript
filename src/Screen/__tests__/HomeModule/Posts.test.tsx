import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { SafeAreaProvider } from 'react-native-safe-area-context';

import Posts from '../../HomeModule/Posts';
import { AppState } from 'src/Redux/Store';

const initialState: AppState = {
	Main: {
		error: null,
		data: {
			posts: [{ id: 1, title: 'title', body: 'body' }],
		},
	},
	_persist: {
		version: -1,
		rehydrated: true,
	},
};

it('render correctly', () => {
	const mockStore = configureStore();
	const store = mockStore(initialState);
	const wrapper = (
		<Provider store={store}>
			<SafeAreaProvider>
				<Posts navigation={null} />
			</SafeAreaProvider>
		</Provider>
	);

	const tree = renderer.create(wrapper).toJSON();

	expect(tree).toMatchSnapshot();
});
