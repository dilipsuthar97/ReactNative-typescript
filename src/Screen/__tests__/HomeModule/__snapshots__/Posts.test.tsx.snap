// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<RNCSafeAreaView
  onInsetsChange={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;
